File: Social/Screens/Perfil/ProfileVisitsViewModel.swift
//
//  ProfileVisitsViewModel.swift
//  Social
//
//  "Quién visitó tu perfil" real, comparado con LinkedIn/Twitter-X
//  (Premium) -- ver ProfileViewerView.swift (registra la visita al abrir
//  un perfil ajeno) y 0132_profile_visits.sql. Mismo patrón sin join
//  embebido que FollowListViewModel.swift: `profile_visits` no trae el
//  perfil del visitante embebido, se resuelve aparte con un solo select.
//  Equivalente de ProfileVisitsViewModel.kt.
//

import Foundation

struct ProfileVisitEntry: Identifiable {
    let id: UUID
    let displayName: String
    let avatarConfig: [String: String]?
    let visitedAt: String
}

@MainActor
final class ProfileVisitsViewModel: ObservableObject {
    @Published var visits: [ProfileVisitEntry] = []
    @Published var errorMessage: String?

    private struct VisitRow: Decodable {
        let visitor_id: UUID
        let visited_at: String
    }

    private struct NameRow: Decodable {
        let id: UUID
        let display_name: String
        let avatar_config: [String: String]?
    }

    func load() async {
        guard let userID = try? await SupabaseManager.shared.client.auth.session.user.id else { return }
        do {
            let rows: [VisitRow] = try await SupabaseManager.shared.client
                .from("profile_visits")
                .select("visitor_id,visited_at")
                .eq("visited_id", value: userID)
                .order("visited_at", ascending: false)
                .limit(100)
                .execute()
                .value
            let visitorIDs = rows.map { $0.visitor_id }
            var byID: [UUID: NameRow] = [:]
            if !visitorIDs.isEmpty {
                let profiles: [NameRow] = try await SupabaseManager.shared.client
                    .from("profiles")
                    .select("id,display_name,avatar_config")
                    .in("id", values: visitorIDs)
                    .execute()
                    .value
                byID = Dictionary(uniqueKeysWithValues: profiles.map { ($0.id, $0) })
            }
            visits = rows.compactMap { row in
                byID[row.visitor_id].map { ProfileVisitEntry(id: row.visitor_id, displayName: $0.display_name, avatarConfig: $0.avatar_config, visitedAt: row.visited_at) }
            }
        } catch {
            errorMessage = "No se pudieron cargar las visitas."
        }
    }
}

File: Android/app/src/main/java/com/social/app/screens/perfil/ProfileVisitsViewModel.kt
package com.social.app.screens.perfil

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.social.app.backend.SupabaseManager
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

data class ProfileVisitEntry(
    val profileId: String,
    val displayName: String,
    val avatarConfig: Map<String, String>? = null,
    val visitedAt: String
)

/**
 * "Quién visitó tu perfil" real, comparado con LinkedIn/Twitter-X
 * (Premium) -- ver ProfileViewerScreen.kt (registra la visita al abrir
 * un perfil ajeno) y 0132_profile_visits.sql. Mismo patrón sin join
 * embebido que FollowListViewModel.kt: `profile_visits` no trae el
 * perfil del visitante embebido, se resuelve aparte con un solo select.
 */
class ProfileVisitsViewModel : ViewModel() {

    private val _visits = MutableStateFlow<List<ProfileVisitEntry>>(emptyList())
    val visits: StateFlow<List<ProfileVisitEntry>> = _visits.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    @Serializable
    private data class VisitRow(
        @SerialName("visitor_id") val visitorId: String,
        @SerialName("visited_at") val visitedAt: String
    )

    @Serializable
    private data class NameRow(
        val id: String,
        @SerialName("display_name") val displayName: String,
        @SerialName("avatar_config") val avatarConfig: Map<String, String>? = null
    )

    fun load() {
        viewModelScope.launch {
            try {
                val myId = SupabaseManager.client.auth.currentUserOrNull()?.id ?: return@launch
                val rows = SupabaseManager.client.from("profile_visits")
                    .select(columns = Columns.raw("visitor_id,visited_at")) {
                        filter { eq("visited_id", myId) }
                        order("visited_at", Order.DESCENDING)
                        limit(100)
                    }
                    .decodeList<VisitRow>()
                val visitorIds = rows.map { it.visitorId }
                val profiles = if (visitorIds.isEmpty()) emptyList() else SupabaseManager.client.from("profiles")
                    .select(columns = Columns.raw("id,display_name,avatar_config")) { filter { isIn("id", visitorIds) } }
                    .decodeList<NameRow>()
                val byId = profiles.associateBy { it.id }
                _visits.value = rows.mapNotNull { row ->
                    byId[row.visitorId]?.let { ProfileVisitEntry(row.visitorId, it.displayName, it.avatarConfig, row.visitedAt) }
                }
            } catch (e: Exception) {
                _errorMessage.value = "No se pudieron cargar las visitas."
            }
        }
    }
}
